zingolib/src/wallet/keys/unified.rs

```diff
@@ -167,52 +183,78 @@ impl WalletCapability {
         {
             return Err("The wallet is not capable of producing desired receivers.".to_string());
         }
-
+        if self
+            .addresses_write_lock
+            .swap(true, atomic::Ordering::Acquire)
+        {
+            return Err("addresses_write_lock collision!".to_string());
+        }
+        let previous_num_addresses = self.addresses.len();
         let orchard_receiver = if desired_receivers.orchard {
-            let fvk: orchard::keys::FullViewingKey = (&*self).try_into()?;
+            let fvk: orchard::keys::FullViewingKey = match self.try_into() {
+                Ok(viewkey) => viewkey,
+                Err(e) => {
+                    self.addresses_write_lock
+                        .swap(false, atomic::Ordering::Release);
+                    return Err(e);
+                }
+            };
             Some(fvk.address_at(self.addresses.len(), Scope::External))
         } else {
             None
         };
 
         // produce a Sapling address to increment Sapling diversifier index
-        let sapling_address = if self.sapling.can_view() {
+        let sapling_receiver = if desired_receivers.sapling && self.sapling.can_view() {
+            let mut sapling_diversifier_index = DiversifierIndex::new();
+            let mut address;
+            let mut count = 0;
             let fvk: zcash_primitives::zip32::sapling::DiversifiableFullViewingKey =
-                (&*self).try_into().unwrap();
-            let (mut new_index, address) = fvk
-                .find_address(self.next_sapling_diversifier_index)
-                .expect("Diversifier index overflow");
-            new_index.increment().expect("Diversifier index overflow");
-            self.next_sapling_diversifier_index = new_index;
+                self.try_into().expect("to create an fvk");
+            loop {
+                (sapling_diversifier_index, address) = fvk
+                    .find_address(sapling_diversifier_index)
+                    .expect("Diversifier index overflow");
+                sapling_diversifier_index
+                    .increment()
+                    .expect("diversifier index overflow");
+                if count == self.addresses.len() {
+                    break;
+                }
+                count += 1;
+            }
```

Comment:
I think a for loop would be cleaner here..something like this?
```suggestion
            for i in 0..self.addresses.len() {
                sapling_diversifier_index
                    .increment()
                    .expect("diversifier index overflow");
            }
            Some(fvk
                .find_address(sapling_diversifier_index)
                .expect("Diversifier index overflow"));
```

Reply:
Did you test this suggestion against the `determinism` test?

Reply:
Talked through this code line-by-line with @fluidvanadium. I revert my suggestion, your fancy loop is nessecary.